File: Backend in .NET/Controllers/AttendanceController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using SmartSchoolAPI.DTOs;
using SmartSchoolAPI.Services;
using System.Security.Claims;

namespace SmartSchoolAPI.Controllers;

[ApiController]
[Route("api/[controller]")]
[Authorize]
public class AttendanceController : ControllerBase
{
    private readonly IAttendanceService _attendanceService;

    public AttendanceController(IAttendanceService attendanceService)
    {
        _attendanceService = attendanceService;
    }

    [HttpPost("mark")]
    [Authorize(Roles = "Teacher,Admin")]
    public async Task<IActionResult> MarkAttendance([FromBody] AttendanceDto attendanceDto)
    {
        var userId = int.Parse(User.FindFirst(ClaimTypes.NameIdentifier)?.Value ?? "0");
        var result = await _attendanceService.MarkAttendanceAsync(attendanceDto, userId);
        if (result)
        {
            return Ok(new { message = "Attendance marked successfully" });
        }
        return BadRequest(new { message = "Failed to mark attendance" });
    }

    [HttpGet("student/{studentId}")]
    public async Task<IActionResult> GetStudentAttendance(int studentId)
    {
        var userId = int.Parse(User.FindFirst(ClaimTypes.NameIdentifier)?.Value ?? "0");
        var role = User.FindFirst(ClaimTypes.Role)?.Value ?? "";

        // Students can only view their own attendance
        if (role == "Student")
        {
            // Verify student is viewing their own attendance
            // This would require additional logic to check studentId matches userId
        }

        var attendance = await _attendanceService.GetStudentAttendanceAsync(studentId);
        return Ok(attendance);
    }

    [HttpGet("all")]
    [Authorize(Roles = "Admin,Teacher")]
    public async Task<IActionResult> GetAllAttendance()
    {
        var attendance = await _attendanceService.GetAllAttendanceAsync();
        return Ok(attendance);
    }

    [HttpGet("{id}")]
    [Authorize(Roles = "Admin")]
    public async Task<IActionResult> GetAttendance(int id)
    {
        var attendance = await _attendanceService.GetAttendanceByIdAsync(id);
        if (attendance == null) return NotFound();
        return Ok(attendance);
    }

    [HttpPut("{id}")]
    [Authorize(Roles = "Admin")]
    public async Task<IActionResult> UpdateAttendance(int id, [FromBody] AttendanceDto attendanceDto)
    {
        var result = await _attendanceService.UpdateAttendanceAsync(id, attendanceDto);
        if (result)
        {
            return Ok(new { message = "Attendance updated successfully" });
        }
        return BadRequest(new { message = "Failed to update attendance" });
    }

    [HttpDelete("{id}")]
    [Authorize(Roles = "Admin")]
    public async Task<IActionResult> DeleteAttendance(int id)
    {
        var result = await _attendanceService.DeleteAttendanceAsync(id);
        if (result)
        {
            return Ok(new { message = "Attendance deleted successfully" });
        }
        return BadRequest(new { message = "Failed to delete attendance" });
    }
}

File: SchoolSync_SpringBoot/backend/src/main/java/com/schoolsync/controller/AttendanceController.java
package com.schoolsync.controller;

import com.schoolsync.dto.AttendanceDTO;
import com.schoolsync.service.AttendanceService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/attendance")
@RequiredArgsConstructor
@CrossOrigin(origins = "*")
public class AttendanceController {

    private final AttendanceService attendanceService;

//    @PostMapping("/mark")
//    public ResponseEntity<?> markAttendance(@RequestBody Map<String, Object> request) {
//        try {
//            if (attendanceService.hasAttendanceBeenMarked(request)) {
//    throw new RuntimeException("Attendance already marked for today.");
//}
//attendanceService.markAttendance(request);
//            return ResponseEntity.ok(Map.of("message", "Attendance marked successfully"));
//        } catch (RuntimeException e) {
//            return ResponseEntity.badRequest().body(Map.of("message", e.getMessage()));
//        }
//    }

    @PostMapping("/mark")
    public ResponseEntity<?> markAttendance(@RequestBody Map<String, Object> request) {
        try {
            // Check if attendance already exists
            boolean exists = attendanceService.checkAttendanceExists(request);
            if (exists) {
                return ResponseEntity.badRequest()
                    .body(Map.of("message", "Attendance already marked for this student on this date"));
            }
            
            attendanceService.markAttendance(request);
            return ResponseEntity.ok(Map.of("message", "Attendance marked successfully"));
        } catch (RuntimeException e) {
            return ResponseEntity.badRequest().body(Map.of("message", e.getMessage()));
        }
    }//
    @GetMapping("/all")
    public ResponseEntity<List<AttendanceDTO>> getAllAttendance() {
        return ResponseEntity.ok(attendanceService.getAllAttendance());
    }

    @GetMapping("/student/{studentId}")
    public ResponseEntity<List<AttendanceDTO>> getStudentAttendance(@PathVariable Long studentId) {
        return ResponseEntity.ok(attendanceService.getStudentAttendance(studentId));
    }

    @PutMapping("/{id}")
    public ResponseEntity<?> updateAttendance(@PathVariable Long id, @RequestBody Map<String, Object> request) {
        try {
            attendanceService.updateAttendance(id, request);
            return ResponseEntity.ok(Map.of("message", "Attendance updated successfully"));
        } catch (RuntimeException e) {
            return ResponseEntity.badRequest().body(Map.of("message", e.getMessage()));
        }
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> deleteAttendance(@PathVariable Long id) {
        try {
            attendanceService.deleteAttendance(id);
            return ResponseEntity.ok(Map.of("message", "Attendance deleted successfully"));
        } catch (RuntimeException e) {
            return ResponseEntity.notFound().build();
        }
    }
}
